pkg/reconciler/namespace/controller.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package namespace
+
+import (
+	"context"
+
+	nsinformer "knative.dev/pkg/injection/informers/kubeinformers/corev1/namespace"
+	kcertinformer "knative.dev/serving/pkg/client/injection/informers/networking/v1alpha1/certificate"
+
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/pkg/configmap"
+	"knative.dev/pkg/controller"
+	"knative.dev/serving/pkg/apis/networking"
+	"knative.dev/serving/pkg/network"
+	"knative.dev/serving/pkg/reconciler"
+	"knative.dev/serving/pkg/reconciler/namespace/config"
+)
+
+const (
+	controllerAgentName = "namespace-controller"
+)
+
+type configStore interface {
+	ToContext(ctx context.Context) context.Context
+	WatchConfigs(w configmap.Watcher)
+}
+
+// NewController initializes the controller and is called by the generated code
+// Registers eventhandlers to enqueue events.
+func NewController(ctx context.Context, cmw configmap.Watcher) *controller.Impl {
+	nsInformer := nsinformer.Get(ctx)
+	knCertificateInformer := kcertinformer.Get(ctx)
+
+	c := &Reconciler{
+		Base:                reconciler.NewBase(ctx, controllerAgentName, cmw),
+		nsLister:            nsInformer.Lister(),
+		knCertificateLister: knCertificateInformer.Lister(),
+	}
+
+	impl := controller.NewImpl(c, c.Logger, "Namespace")
+
+	c.Logger.Info("Setting up event handlers")
+	nsInformer.Informer().AddEventHandler(cache.FilteringResourceEventHandler{
+		FilterFunc: reconciler.LabelExistsFilterFunc(networking.WildcardCertLabelKey),
```

Comment:
So namespaces must be labeled to get wildcard certs?  I'd rather have users opt-out per-namespace once they have opted in per-cluster.

Reply:
cc @ZhiminXiang (I may be missing some context or misinterpreting).

Reply:
I had it as opt-out originally and changed it so that certs for system namespaces wouldn't get created as soon as the controller started. Having to opt-in per cluster mitigates that. Users will just have to remember to opt-out prior to enabling the controller.

Reply:
I think we want to have opt-out for at least all user-defined namespace as the goal of the this controller is to "eagerly" provide certificate when namespace is created.

For system namespace, I think it is probably OK to also create certs for them although they may not be actually used. Or we can make opt-in for system namespaces, i.e. for system namespaces such as `kube-system`, `knative-serving`, we only create certs when it explicitly has label "wildcarcert:true".

WDYT?

Reply:
I think it's better to just let them be created if those namespaces aren't excluded. Having some namespaces be opt-out and some opt-in sounds confusing.